identity-service/src/relay/ethTxRelay.js

```diff
@@ -0,0 +1,213 @@
+
+const ENVIRONMENT = config.get('environment')
+const DEFAULT_GAS_LIMIT = config.get('defaultGasLimit')
+const GANACHE_GAS_PRICE = config.get('ganacheGasPrice')
+
+// L1 relayer wallets
+let ethRelayerWallets = [...ethRelayerConfigs] // will be array of { locked, publicKey, privateKey }
+ethRelayerWallets.forEach(wallet => {
+  wallet.locked = false
+})
+
+async function delay (ms) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+// Calculates index into eth relayer addresses
+const getEthRelayerWalletIndex = (walletAddress) => {
+  let walletParsedInteger = parseInt(walletAddress, 16)
+  return walletParsedInteger % ethRelayerWallets.length
+}
+
+// Select from the list of eth relay wallet addresses
+// Return the public key that will be used to relay this address
+const queryEthRelayerWallet = (walletAddress) => {
+  return ethRelayerWallets[getEthRelayerWalletIndex(walletAddress)].publicKey
+}
+
+// Query current balance for a given relayer public key
+const getEthRelayerFunds = async (walletPublicKey) => {
+  return ethWeb3.eth.getBalance(walletPublicKey)
+}
+
+const selectEthWallet = async (walletPublicKey, reqLogger) => {
+  reqLogger.info(`Acquiring lock for ${walletPublicKey}`)
+  let ethWalletIndex = getEthRelayerWalletIndex(walletPublicKey)
+  while (ethRelayerWallets[ethWalletIndex].locked) {
+    await delay(200)
+  }
+  ethRelayerWallets[ethWalletIndex].locked = true
+  reqLogger.info(`Locking ${ethRelayerWallets[ethWalletIndex].publicKey}, index=${ethWalletIndex}}`)
+  return {
+    selectedEthRelayerWallet: ethRelayerWallets[ethWalletIndex],
+    ethWalletIndex
+  }
+}
+
+// Relay a transaction to the ethereum network
+const sendEthTransaction = async (req, txProps, reqBodySHA, onTxHash) => {
+  const {
+    contractAddress,
+    encodedABI,
+    senderAddress,
+    gasLimit
+  } = txProps
+
+  // Calculate relayer from senderAddress
+  let { selectedEthRelayerWallet, ethWalletIndex } = await selectEthWallet(senderAddress, logger)
+  req.logger.info(`L1 txRelay - selected relayerPublicWallet=${selectedEthRelayerWallet.publicKey}`)
+  let ethGasPriceInfo = await getProdGasInfo(req.app.get('redis'), req.logger)
+
+  // Select the 'fast' gas price
+  let ethRelayGasPrice = ethGasPriceInfo.fastGweiHex
+  let txHash
+  try {
+    txHash = await createAndSendEthTransaction(
+      {
+        publicKey: selectedEthRelayerWallet.publicKey,
+        privateKey: selectedEthRelayerWallet.privateKey
+      },
+      contractAddress,
+      '0x00',
+      ethWeb3,
+      req.logger,
+      ethRelayGasPrice,
+      onTxHash,
+      gasLimit,
+      encodedABI
+    )
+  } catch (e) {
+    req.logger.error('L1 txRelay - Error in relay', e)
+  } finally {
+    req.logger.info(`L1 txRelay - Unlocking ${ethRelayerWallets[ethWalletIndex].publicKey}, index=${ethWalletIndex}}`)
+    // Unlock wallet
+    ethRelayerWallets[ethWalletIndex].locked = false
+  }
+
+  selectedEthRelayerWallet.selectedRelay = selectedEthRelayerWallet
```

Comment:
What's this line for?

Reply:
this used to be added in the txReceipt info, i guess is unneeded now